Source/DafnyLanguageServer/Workspace/Compilation.cs

```diff
@@ -0,0 +1,175 @@
+﻿using System;
+using OmniSharp.Extensions.LanguageServer.Protocol;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Boogie;
+using Microsoft.Dafny.LanguageServer.Language;
+using Microsoft.Dafny.LanguageServer.Workspace.ChangeProcessors;
+using SymbolTable = Microsoft.Dafny.LanguageServer.Language.Symbols.SymbolTable;
+using Microsoft.Dafny.LanguageServer.Workspace.Notifications;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Range = OmniSharp.Extensions.LanguageServer.Protocol.Models.Range;
+
+namespace Microsoft.Dafny.LanguageServer.Workspace {
+
+  /// <summary>
+  /// Internal representation of a specific version of a Dafny document.
+  ///
+  /// Only one instance should exist of a specific version.
+  /// Asynchronous compilation tasks use this instance to synchronise on
+  ///
+  /// When verification starts, no new instances of Compilation will be created for this version.
+  /// There can be different verification threads that update the state of this object.
+  /// </summary>
+  public class Compilation {
+    public DocumentTextBuffer TextDocumentItem { get; }
+    public DocumentUri Uri => TextDocumentItem.Uri;
+    public int Version => TextDocumentItem.Version!.Value;
+
+    public Compilation(DocumentTextBuffer textDocumentItem) {
+      TextDocumentItem = textDocumentItem;
+    }
+
+    public virtual IEnumerable<Diagnostic> Diagnostics => Enumerable.Empty<Diagnostic>();
+
+    public CompilationView NotMigratedSnapshot() {
+      return Snapshot(new CompilationView(TextDocumentItem, Array.Empty<Diagnostic>(),
+        SymbolTable.Empty(TextDocumentItem), new Dictionary<ImplementationId, ImplementationView>(),
+        false, Array.Empty<Diagnostic>(),
+        new DocumentVerificationTree(TextDocumentItem)));
+    }
+
+    /// <summary>
+    /// Creates a clone of the DafnyDocument
+    /// </summary>
+    public virtual CompilationView Snapshot(CompilationView previousView) {
```

Comment:
Since all your method named `Snapshot` return a `CompilationView`, I suggest you choose between `View` and `Snapshot`. I prefer `Snapshot` because it's less ambiguous than `View` (a view of a data structure can change if the underlying data structure change), but I understand that verification trees are updated anyway so `View` might be preferred.